torchtitan/checkpoint.py

```diff
@@ -220,44 +176,29 @@ def __init__(
 
             TODO: This is currently unsolved and needs a fix.
         """
-        assert len(model_parts) == len(
+        if job_config.optimizer.early_step_in_backward:
+            expected_optim_num = sum(
+                len([param for param in model.parameters()]) for model in model_parts
+            )
+        else:
+            expected_optim_num = len(model_parts)
+        assert expected_optim_num == len(
             optimizers.optimizers
-        ), "Must pass one optimizer per model part"
-        assert len(model_parts) == len(
+        ), "Must pass one optimizer per model part (or per param for optim_in_bwd)"
+        assert expected_optim_num == len(
             lr_schedulers.schedulers
-        ), "Must pass one lr_scheduler per model part"
+        ), "Must pass one lr_scheduler per model part (or per param for optim_in_bwd)"
 
         self.states = states
 
         self.states.update(
             {
                 "model": ModelWrapper(model_parts),
-                "optimizer": OptimizerWrapper(
-                    model_parts,
-                    optimizers,
-                ),
+                "optimizer": optimizers,
                 "dataloader": dataloader,
             }
         )
-        # SchedulersInBackwardContainer has a different structure than SchedulersContainer, List[List[Scheduler]] rahter
-        # than List[Scheduler], but the schedulers are the same for each list inside, so here just store the first one.
-        # TODO: Restructure SchedulersInBackwardContainer to be consisitent with SchedulersContainer.
-        if isinstance(lr_schedulers, SchedulersInBackwardContainer):
-            if len(lr_schedulers.schedulers) == 1:
-                self.states["lr_scheduler"] = lr_schedulers.schedulers[0][0]
-            else:
-                # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
-                # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
-                for idx, lr_scheduler in enumerate(lr_schedulers.schedulers):
-                    self.states[f"lr_scheduler_{idx}"] = lr_scheduler[0]
-        else:
-            if len(lr_schedulers.schedulers) == 1:
-                self.states["lr_scheduler"] = lr_schedulers.schedulers[0]
-            else:
-                # For now, pipeline-parallel with looped schedules does not support resharding for lr_scheduler.
-                # It should only support saving and loading a distributed checkpoint with the same number of pp ranks
-                for idx, lr_scheduler in enumerate(lr_schedulers.schedulers):
-                    self.states[f"lr_scheduler_{idx}"] = lr_scheduler
+        self.states.update(lr_schedulers.get_lr_scheduler_state())
```

Comment:
Discussed with @mori360 offline: I think as a next step, we should still have a single entry for `lr_scheduler`. To achieve that, we need to understand if any further flattening is needed. This could potentially solve the "PP multi-schedule doesn't support DCP resharding" problem.